rclcpp/src/rclcpp/utilities.cpp

```diff
@@ -42,116 +43,233 @@
 #define HAS_SIGACTION
 #endif
 
-/// Mutex to protect installation and uninstallation of signal handlers.
-static std::mutex g_signal_handlers_mutex;
-/// Atomic bool to control setup of signal handlers.
-static std::atomic<bool> g_signal_handlers_installed(false);
+namespace rclcpp
+{
 
-/// Mutex to protect g_sigint_guard_cond_handles
-static std::mutex g_sigint_guard_cond_handles_mutex;
-/// Guard conditions for interrupting the rmw implementation when the global interrupt signal fired.
-static std::map<rcl_wait_set_t *, rcl_guard_condition_t> g_sigint_guard_cond_handles;
+class SignalHandler final
+{
+public:
+  static
+  SignalHandler &
+  get_global_signal_handler()
+  {
+    static SignalHandler singleton;
+    return singleton;
+  }
 
-/// Condition variable for timed sleep (see sleep_for).
-static std::condition_variable g_interrupt_condition_variable;
-/// Mutex for protecting the global condition variable.
-static std::mutex g_interrupt_mutex;
+  bool
+  install()
+  {
+    bool already_installed = installed_.exchange(true);
+    if (already_installed) {
+      return false;
+    }
+    try {
+      signal_received.exchange(false);
+#ifdef HAS_SIGACTION
+      struct sigaction action;
+      memset(&action, 0, sizeof(action));
+      sigemptyset(&action.sa_mask);
+      action.sa_sigaction = signal_handler;
+      action.sa_flags = SA_SIGINFO;
+      old_action = set_sigaction(SIGINT, action);
+#else
+      old_signal_handler = set_signal_handler(SIGINT, signal_handler);
+#endif
+      signal_handler_thread_ =
+        std::thread(&SignalHandler::deferred_signal_handler, this);
+    } catch (...) {
+      installed_.exchange(false);
+      throw;
+    }
+    RCLCPP_DEBUG(rclcpp::get_logger("rclcpp"), "signal handler installed");
+    return true;
+  }
 
+  bool
+  uninstall()
+  {
+    bool installed = installed_.exchange(false);
```

Comment:
Is the goal to make `install()` and `uninstall()` thread-safe with each other?

If so, starting from signal handlers uninstalled, it looks like thread1 calling `install()` could set `installed_` to `true`, context switch to thread2 which calls `uninstall()` that sets `installed_` to `false` and completes, then switch back to thread1 which installs the signal handler. The end result is the signal handler is installed but `is_installed()` would return `false`.

Reply:
that sounds correct to me. I guess these two functions have to be locked by a mutex then.